lib/model/message_list.dart

```diff
@@ -86,6 +88,67 @@ class MessageListView extends ChangeNotifier {
     notifyListeners();
   }
 
+  applyChangesToMessage(UpdateMessageEvent event, Message message) {
+    if (event.renderingOnly != null && event.renderingOnly == true) {
+      //TODO update inline preview
```

Comment:
Assuming this is out of scope for this issue.

Reply:
Yeah, after reading that doc, this is what I thought it was describing:

1. User sends a message with a link
2. Server sends/client receives `MessageEvent`, with content like `<p>My message: <a href="somewhere.com">Link</a>`
3. Client adds the message
4. Server finishes rendering a preview of `somewhere.com`
5. Server sends/client receives `UpdateMessageEvent`, with `renderingOnly` == `true`, and content like `<p>My message: <a href="somewhere.com">Link</a></p> <div className="websitePreview"><img src= ...`
6. Client (potentially) needs to handle that message content differently, due to the extra link preview content, hence the `//todo`

I tried to DM myself a few links, but didn't get any inline URL previews in them, so I was unable to test that assumption.

Reply:
> I tried to DM myself a few links, but didn't get any inline URL previews in them, so I was unable to test that assumption.

Hmm, yeah, we should interlink these docs a bit more. There's an organization-level setting that controls that feature:
https://zulip.com/help/allow-image-link-previews
and it is off on chat.zulip.org. So to exercise the feature, one needs a test organization elsewhere — e.g. from a local dev server, or from making a test organization on zulipchat.com. Or you could DM yourself in another existing organization if that setting is enabled. (To check if it is, you can always view most organization-level settings even when you don't have permission to set them.)

---

In your hypothesis, steps 1 to 5 are correct. Then in step 6, it's true that the parsing and rendering of the message content will need to handle the link preview. But that's independent of the fact that the client saw the message get updated in this way: if the user goes to browse some messages and there's an inline URL preview already there, then that'll have the same sort of content and need the same parsing and rendering logic to handle it. Conversely, I think inline URL previews may currently be the only case where the server sends these "rendering only" updates, but the API doesn't promise that — there may be other such cases, now or in the future.

For this code that's handling `UpdateMessageEvent`, its job in general is to get things to the same state as we'd get if we went and fetched the message from scratch. So that's what it should do in this case too.

Reply:
BTW, I think our parser doesn't currently support inline URL previews, but that's fine. That is, although we'll want to implement it at some point, it won't make anything fall over in the meantime.

That's because the parsing and rendering is designed to be robust to the existence of unimplemented features in the message content. If you scroll through history on chat.zulip.org, you can usually find at least one message that has a red "(unimplemented: …)" marker; or the quickest way to get such a thing in a test message is perhaps:
```markdown
# a header
```

See the `_errorUnimplemented` function in widgets/content.dart for more about that, and #190 for our related plans.

Reply:
Aaah, I see - so just to make sure I understand correctly, the desired behavior for step 6 here then is that we set `message.renderedContent` to the new content, in the same way as we do for a normal "user-updated" message? But ignoring the timestamp, because it should have been unchanged?

Aha, I have seen some of those 'unimplemented' elements go by and was wondering about that. Mystery solved!

Reply:
> Aaah, I see - so just to make sure I understand correctly, the desired behavior for step 6 here then is that we set `message.renderedContent` to the new content, in the same way as we do for a normal "user-updated" message?

Right, exactly.


> But ignoring the timestamp, because it should have been unchanged?

Yeah. Specifically the event will have an `edit_timestamp` (except on old servers) and it'll be a fresh recent time… but the edit isn't one initiated by a user, so we'll leave it out of any record of the message having been edited — effectively we'll let the server make a stealth edit.



> Aha, I have seen some of those 'unimplemented' elements go by and was wondering about that. Mystery solved!

:slightly_smiling_face:

Reply:
Implemented and tested!